### PR TITLE
doc: config_and_build: regroup and rework pages

### PR DESCRIPTION
**Pull request by greg-fer**
Moved sections about CMake and Kconfig from modifying.rst to separate pages and removed modifying.rst.
Split Building and programming into two separate sections. Grouped guides for configuring under Configuring and building the app. NCSDK-22324.

----

- [x] Add redirects
- [x] Add tags

There will be one more PR for final cleanup of these new pages.

### CODE REVIEW

File: doc/nrf/config_and_build.rst [hunk 1]
```diff
@@ -4,24 +4,27 @@
 Configuration and building
 ##########################
 
-The |NCS| build and configuration system is based on the one from :ref:`Zephyr project <zephyr:getting_started>`, with some additions.
+After you :ref:`created an application <create_application>`, you need to configure and build it in order to be able to run it.
```
Comment (nordicjm) [inline doc/nrf/config_and_build.rst:7]:
```suggestion
After you have :ref:`created an application <create_application>`, you need to configure and build it in order to be able to run it.
```

File: doc/nrf/config_and_build.rst [hunk 2]
```diff
@@ -4,24 +4,27 @@
...
 
-The figure below visualizes the tools and configuration methods in the |NCS|.
+The |NCS| configuration and build system is based on the one from :ref:`Zephyr project <zephyr:getting_started>`, with some additions.
```
Comment (nordicjm) [inline doc/nrf/config_and_build.rst:9]:
```suggestion
The |NCS| configuration and build system is based on the one from :ref:`Zephyr <zephyr:getting_started>`, with some additions.
```

File: doc/nrf/config_and_build/configuring_app/advanced_building.rst [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+.. _modifying_build_types:
+
+Configuring build types
+***********************
+
+.. build_types_overview_start
+
+When the ``CONF_FILE`` variable contains a single file and this file follows the naming pattern :file:`prj_<buildtype>.conf`, then the build type will be inferred to be *<buildtype>*.
+The build type cannot be set explicitly.
+The *<buildtype>* can be any string, but it is common to use ``release`` and ``debug``.
```
Comment (nordicjm) [inline doc/nrf/config_and_build/configuring_app/advanced_building.rst:22]: We should probably not be documenting this because I'm hoping to dump this feature soon and have it replaced with something universal

File: doc/nrf/config_and_build/configuring_app/cmake/index.rst [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+Adding source files to CMakeLists.txt
+*************************************
+
+You can add source files to the ``app`` CMake target with the :c:func:`target_sources` function provided by CMake.
+
+Pay attention to the following configuration options:
+
+* If your application is complex, you can split it into subdirectories.
+  These subdirectories can provide their own :file:`CMakeLists.txt` files.
```
Comment (nordicjm) [inline doc/nrf/config_and_build/configuring_app/cmake/index.rst:30]: the main file would need to include these

File: doc/nrf/config_and_build/configuring_app/index.rst [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+Configuring and building an application
+#######################################
+
+.. contents::
+   :local:
+   :depth: 2
+
+|application_sample_definition|
+
+After you :ref:`created an application <create_application>`, you need to build it in order to be able to program it.
```
Comment (nordicjm) [inline doc/nrf/config_and_build/configuring_app/index.rst:14]:
```suggestion
After you have :ref:`created an application <create_application>`, you need to build it in order to be able to program it.
```

File: doc/nrf/config_and_build/configuring_app/advanced_building.rst (hunk 3 above)
Comment (greg-fer) [reply doc/nrf/config_and_build/configuring_app/advanced_building.rst:22]: I would not want to remove it here, as this affects a lot of pages for apps and samples. Let's review the docs when the feature is being replaced. When do you think this might happen?

File: doc/nrf/config_and_build/configuring_app/advanced_building.rst [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+For information about how to set variables, see :ref:`zephyr:important-build-vars` in the Zephyr documentation.
+
+The following software components can be made dependent on the build type:
+
+* The Partition Manager's :ref:`static configuration <ug_pm_static>`.
+  When the build type has been inferred, the file :file:`pm_static_<buildtype>.yml` will have precedence over :file:`pm_static.yml`.
+* The :ref:`child image Kconfig configuration <ug_multi_image_permanent_changes>`.
+  The child image Kconfig overlay file is named :file:`child_image/<child_image_name>.conf` instead of :file:`prj.conf`, but otherwise follows the same pattern as the parent Kconfig.
```
Comment (Damian-Nordic) [inline doc/nrf/config_and_build/configuring_app/advanced_building.rst:31]: Don't quite understand why child image Kconfig overlay is compared to `prj.conf` as these are two different things.
How about something like:
When the build type has been inferred, the child image Kconfig overlay file is searched at :file:`child_image/<child_image_name>_<buildtype>.conf`

Btw, given how complex child image feature is maybe it's better not repeat partial information from the multi image page and just state that "certain child image configuration files located at :file:`child_image/` directory can be defined per build type.

File: doc/nrf/config_and_build/programming.rst [hunk 7]
```diff
@@ -1,133 +1,88 @@
...
-  For example, if you run the west build command with an additional parameter ``@1.0.0`` for nRF9160 build target, it adds the external flash on the nRF9160 DK that was available since board version 0.14.0.
-* You can :ref:`start menuconfig with the west command <configuration_temporary_change>` to configure your application.
-* You can reuse an existing build directory for building another application for another board or build target by passing ``-p=auto`` to ``west build``.
+  .. code-block:: console
 
-For more information on other optional build parameters, run the ``west build -h`` help text command.
+     west flash
 
-.. |output_files_note| replace:: For more information about files generated as output of the build process, see :ref:`app_build_output_files`.
+Programming the nRF52840 Dongle
```
Comment (Damian-Nordic) [inline doc/nrf/config_and_build/programming.rst:87]: This is not really optional programming parameter. More suits to "Hardware-specific programming steps" section.

File: doc/nrf/config_and_build/configuring_app/advanced_building.rst [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+* Some applications contain configuration overlay files that enable specific features.
+  These can be added to the ``west build`` command as follows:
+
+  .. parsed-literal::
+     :class: highlight
+
+     west build -b *build_target* -- -DOVERLAY_CONFIG="overlay-feature1.conf;overlay-feature2.conf"
+
+  See :ref:`configuration_permanent_change` and Zephyr's :ref:`zephyr:west-building-cmake-args` for more information.
+* You can include the *directory_name* parameter to build from a directory other than the application directory.
```
Comment (Damian-Nordic) [inline doc/nrf/config_and_build/configuring_app/advanced_building.rst:100]:
```suggestion
* You can include the *directory_name* parameter to build from a directory other than the current directory.
```
Does this mean the source directory? If so, let's call it like that and provide an example. Source directory must be put before `--` so it may not be easy to understand without an example.

File: doc/nrf/config_and_build/configuring_app/cmake/index.rst [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+   .. group-tab:: Command line
+
+      If you work on the command line, pass the additional options to the ``west build`` command when :ref:`building`.
+      The options must be added after a ``--`` at the end of the command.
+      See :ref:`zephyr:west-building-cmake-args` for more information.
+
+.. _compiler_settings:
+
+Advanced compiler settings
```
Comment (Damian-Nordic) [inline doc/nrf/config_and_build/configuring_app/cmake/index.rst:71]: This section is not directly related to CMake. Maybe this could be moved to `advanced_building`?

File: doc/nrf/config_and_build/configuring_app/advanced_building.rst (hunk 6 above)
Comment (greg-fer) [reply doc/nrf/config_and_build/configuring_app/advanced_building.rst:31]: Both build types and child images are going to be reworked soon by @nordicjm , so this is not a critical edit, but I opted for option 2 and also moved some info to multi-image docs.

File: doc/nrf/config_and_build/config_and_build_system.rst [hunk 10]
```diff
@@ -149,10 +154,10 @@ Building phase
 ==============
 
 During this phase, the final build scripts are executed.
-The build phase begins when the user invokes ``make`` or ``ninja``.
+The build phase begins when the user invokes ``make`` or `ninja <Ninja documentation_>`_.
 You can customize this stage by :ref:`cmake_options`.
 
-The result of this process is a complete application in a format suitable for flashing on the desired target board.
+The result of this process is a complete application in a format suitable for flashing on the desired target board, created by the compiler (for example, `GCC compiler`_).
```
[the remark this answers is not in the document]
Comment (greg-fer) [reply doc/nrf/config_and_build/config_and_build_system.rst:160]: Removed the edit for now. Will clarify in the final PR.

**Review by Damian-Nordic — APPROVED** [review]

File: doc/nrf/config_and_build/programming.rst [hunk 11]
```diff
@@ -1,133 +1,88 @@
...
-
-      If you have installed the |NCS| using the :ref:`legacy Toolchain Manager <gs_assistant>`, click the down arrow next to the version you installed and select :guilabel:`Open bash`.
-
-      .. figure:: ../installation/images/gs-assistant_tm_dropdown.png
-         :alt: The Toolchain Manager dropdown menu for the installed nRF Connect SDK version, cropped
-
-         The Toolchain Manager dropdown menu options
-
-#.    Go to the specific application directory.
+To program the :ref:`output build files <app_build_output_files>` to your device, use the steps for the development environment you have been using for the whole project.
```
Comment (hellesvik-nordic) [inline doc/nrf/config_and_build/programming.rst:15]: "using for the whole project" is unspecific, and I don't quite understand what it means. Do the sentence suffer if we just cut it?
```suggestion
To program the :ref:`output build files <app_build_output_files>` to your device, use the steps for the development environment.
```

Comment (hellesvik-nordic) [inline doc/nrf/config_and_build/programming.rst:15]: Also, the amount of build files in Output Build Files can be a bit overwhelming. Can I suggest that we state that zephyr.hex/merged.hex are the ones used by west flash by default here?

File: doc/nrf/config_and_build/programming.rst [hunk 12]
```diff
@@ -1,133 +1,88 @@
...
-         cd nrf/samples/cellular/at_client
+      You need to set up a build configuration before you program a device in the |nRFVSC|.
 
-#.    Build the application using the west command.
-      The build target is specified by the parameter *build_target* in the west command as follows:
+      Complete the following steps:
 
-      .. parsed-literal::
-         :class: highlight
+      1. Connect your device to the terminal.
```
Comment (hellesvik-nordic) [inline doc/nrf/config_and_build/programming.rst:25]: Why does this say "Connect your device to the terminal", while the Command Line tab says "Connect the development kit to your PC using a USB cable.". Can I suggest that both these say the same, to be consistent? I like the latter phrasing better myself.

Comment (hellesvik-nordic) [reply doc/nrf/config_and_build/programming.rst:25]: Same with "Power on the device" and "Hardware specific steps". Any reason this is not in both?

File: doc/nrf/config_and_build/programming.rst [hunk 13]
```diff
@@ -1,133 +1,88 @@
...
-         west flash
+Some hardware has platform-specific requirements that must be met before you program the device.
+For additional information, check the :ref:`user guide for the hardware platform <device_guides>` that you are using.
 
-      The ``west flash`` command automatically resets the kit and starts the application.
+For example, if you are working with an nRF9160 DK, you need to select the correct position of the :ref:`board controller <nrf9160_ug_intro>` switch **SW10** before you program the application to your development kit.
+Programming to Thingy:91 also requires a :ref:`similar step <building_pgming>`, but using a different switch (**SW2**).
 
-For more information on building and programming using the command line, see the Zephyr documentation on :ref:`zephyr:west-build-flash-debug`.
+Programming the nRF52840 Dongle
```
Comment (hellesvik-nordic) [inline doc/nrf/config_and_build/programming.rst:68]: Do we have any docs for programming custom boards somewhere? If so, that would be relevant as an additional point to make in this section.

File: doc/nrf/config_and_build/programming.rst [hunk 14]
```diff
@@ -1,133 +1,88 @@
...
-  For example, if you run the west build command with an additional parameter ``@1.0.0`` for nRF9160 build target, it adds the external flash on the nRF9160 DK that was available since board version 0.14.0.
-* You can :ref:`start menuconfig with the west command <configuration_temporary_change>` to configure your application.
-* You can reuse an existing build directory for building another application for another board or build target by passing ``-p=auto`` to ``west build``.
+  To erase only the areas of flash memory that are required for programming the new application, use the following command:
 
-For more information on other optional build parameters, run the ``west build -h`` help text command.
+  .. code-block:: console
 
-.. |output_files_note| replace:: For more information about files generated as output of the build process, see :ref:`app_build_output_files`.
+     west flash
```
Comment (hellesvik-nordic) [inline doc/nrf/config_and_build/programming.rst:88]: I also often use "west flash --recover".
Since "west flash --help" is not very useful, as it lists generic west options and not specific ones for the nRF series, I think we should list all west flash options we support, such as --erase and --recover. I do not know if there are more than that or not. Maybe this would be a good place to list them?

File: doc/nrf/config_and_build/programming.rst (hunk 12 above)
Comment (greg-fer) [reply doc/nrf/config_and_build/programming.rst:25]: That's because the tech writer was low on coffee. ¯\_(ツ)_/¯ Thanks for the suggestion. Will apply!

File: doc/nrf/config_and_build/configuring_app/kconfig/index.rst [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+   It does not adjust visible symbols that have already defaulted to a value even if this default calculation is supposed to be dependent on the changed setting.
+   This may result in a bloated configuration compared to changing the setting directly in :file:`prj.conf`.
+   See the section Stuck symbols in menuconfig and guiconfig on the :ref:`kconfig_tips_and_tricks` in the Zephyr documentation for more information.
+
+.. tabs::
+
+   .. group-tab:: nRF Connect for VS Code
+
+      If you work with |nRFVSC|, you can use one of the following options:
+
```
[the remark this answers is not in the document]
Comment (greg-fer) [reply doc/nrf/config_and_build/configuring_app/kconfig/index.rst:78]: This could be understood as editing the file directly, which we don't want. Edited a line below instead.

File: doc/nrf/config_and_build/configuring_app/kconfig/index.rst [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+      * Select an extra Kconfig fragment file when you `build an application <How to build an application_>`_.
+      * Edit the Kconfig options using the nRF Kconfig GUI and save changes permanently to an existing or new :file:`prj.conf` file.
+
+      See the `extension's documentation about Kconfig <Configuring with nRF Kconfig_>`_ for more information.
+
+   .. group-tab:: Command line
+
+      If you work on the command line, pass the additional options to the ``west build`` command.
```
[the remark this answers is not in the document]
Comment (greg-fer) [reply doc/nrf/config_and_build/configuring_app/kconfig/index.rst:86]: Made some edits. I will make more in the last PR to this section.

File: doc/nrf/config_and_build/configuring_app/kconfig/index.rst [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+      The options must be added after a ``--`` at the end of the command.
+      See :ref:`zephyr:west-building-cmake-config` for more information.
+
+The :file:`prj.conf` file is read when you open a project.
```
[the remark this answers is not in the document]
Comment (greg-fer) [reply doc/nrf/config_and_build/configuring_app/kconfig/index.rst:90]: You can also open a project in command line if you open a dir and run `code .`

File: doc/nrf/config_and_build/programming.rst (hunk 13 above)
Comment (greg-fer) [reply doc/nrf/config_and_build/programming.rst:68]: No, we need them. I'll handle this update separately.

**Review by divipillai — COMMENTED** [review]
Looks good to me. Some minor suggestions.

File: doc/nrf/config_and_build/programming.rst (hunk 14 above)
Comment (greg-fer) [reply doc/nrf/config_and_build/programming.rst:88]: Erase is documented in the lines 44-52 as the default procedure.

File: doc/nrf/config_and_build/configuring_app/kconfig/index.rst [hunk 18]
```diff
@@ -0,0 +1,122 @@
+.. _configuring_kconfig:
+
+Configuring Kconfig
```
[the remark this answers is not in the document]
Comment (greg-fer) [reply doc/nrf/config_and_build/configuring_app/kconfig/index.rst:3]: It's already there, but in the wrong section. Applied.

File: doc/nrf/config_and_build/configuring_app/kconfig/index.rst (hunk 15 above)
[the remark this answers is not in the document]
Comment (greg-fer) [reply doc/nrf/config_and_build/configuring_app/kconfig/index.rst:79]: Edited after follow-up.

File: doc/nrf/config_and_build/configuring_app/kconfig/index.rst [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+      .. code-block:: console
+
+         west build -t menuconfig
+
+      See :ref:`zephyr:menuconfig` in the Zephyr documentation for instructions on how to run menuconfig or guiconfig.
+      To locate a specific configuration option, use the **Jump to** field.
+
+.. _configuration_permanent_change:
+
+Permanent Kconfig changes
```
[the remark this answers is not in the document]
Comment (greg-fer) [reply doc/nrf/config_and_build/configuring_app/kconfig/index.rst:59]: Edited after follow-up.

File: doc/nrf/config_and_build/configuring_app/kconfig/index.rst (hunk 17 above)
[the remark this answers is not in the document]
Comment (greg-fer) [reply doc/nrf/config_and_build/configuring_app/kconfig/index.rst]: Simplified this paragraph.

File: doc/nrf/config_and_build/programming.rst (hunk 12 above)
Comment (greg-fer) [reply doc/nrf/config_and_build/programming.rst]: Edited now.

File: doc/nrf/config_and_build/configuring_app/kconfig/index.rst [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+###################
+
+.. contents::
+   :local:
+   :depth: 2
+
+You can change the software-related configuration temporarily or permanently.
+The temporary build files are deleted when you clean the build directory with the ``pristine`` target (see Zephyr's :ref:`zephyr:application_rebuild` for more information).
+
+.. note::
```
[the remark this answers is not in the document]
Comment (greg-fer) [reply doc/nrf/config_and_build/configuring_app/kconfig/index.rst:13]: Moved.

File: doc/nrf/images/ncs-toolchain.svg
Comment (umapraseeda) [inline doc/nrf/images/ncs-toolchain.svg:1]: Few observations on the image after consulting @mirjaku, 
1. CMakelists.txt block and the dotted line could have a slightly stronger outline color.
2. The lines denoting the phases could be lines with bidirectional arrows to indicate the phases clearly.

File: doc/nrf/config_and_build/configuring_app/index.rst [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+
+|application_sample_definition|
+
+After you have :ref:`created an application <create_application>`, you need to build it in order to be able to program it.
+Just as for creating the application, you can build the application using either the |nRFVSC| or the command line.
+
+.. note::
+    |application_sample_long_path_windows|
+
+.. tabs::
```
[the remark this answers is not in the document]
Comment (greg-fer) [reply doc/nrf/config_and_build/configuring_app/index.rst:20]: I want this page to provide meaningful information, and in the context of this section, the most meaningful and immediate information is about how to build an app.

If the building steps are moved to a page of its own, the page would end up somewhere in the middle of subpages, where this info would not be prominent within the section. If it is on the advanced building options, then the page would again mix the bare building instructions with more advanced building instructions, resulting in long, convoluted content. Not to mention that all the subpages are at some point mentioning how to customize the building instruction, so placing there really isn't a better place for placing building instructions than on the landing page.

File: doc/nrf/config_and_build/programming.rst [hunk 22]
```diff
@@ -1,133 +1,104 @@
...
-         :alt: The Toolchain Manager dropdown menu for the installed nRF Connect SDK version, cropped
+      You need to set up a build configuration before you program a device in the |nRFVSC|.
 
-         The Toolchain Manager dropdown menu options
+      Complete the following steps:
 
-#.    Go to the specific application directory.
+      1. Connect the development kit to your PC using a USB cable.
+      #. Power on the development kit.
+      #. Make sure you are familiar with :ref:`programming_hw`.
```
[the remark this answers is not in the document]
Comment (greg-fer) [reply doc/nrf/config_and_build/programming.rst:27]: I don't think so. I can rename the section to requirements, but the page should give the users information about the basic programming procedures immediately. On this page, the type of hardware one is using is of secondary importance. The goal is to describe NCS' main way to program.

File: doc/nrf/config_and_build/programming.rst [hunk 23]
```diff
@@ -1,133 +1,104 @@
...
+      #. Open the nRF Connect extension in |VSC| by clicking its icon in the :guilabel:`Activity Bar`.
+      #. In the extension's :guilabel:`Actions View`, click on :guilabel:`Flash`.
```
[the remark this answers is not in the document]
Comment (greg-fer) [reply doc/nrf/config_and_build/programming.rst:29]: "on" is acceptable here based on the criteria of Merriam-Webster we're referencing in Terms.

File: doc/nrf/images/ncs-toolchain.svg
Comment (greg-fer) [reply doc/nrf/images/ncs-toolchain.svg:1]: 1. The outline lines are matching the CMake block color, and I think there is some value in keeping them consistent.
2. I think I found a better solution, will push it shortly.

Comment (umapraseeda) [reply doc/nrf/images/ncs-toolchain.svg:1]: Yes, good to have the CMake block color, probably a slightly stronger shade will work here.

File: doc/nrf/config_and_build/configuring_app/index.rst (hunk 21 above)
[the remark this answers is not in the document]
Comment (umapraseeda) [reply doc/nrf/config_and_build/configuring_app/index.rst:20]: Discussed offline

**Comment by greg-fer** [comment]
@umapraseeda , edits applied. Kindly please check also tags and redirects, which I've just pushed.

**Review by divipillai — APPROVED** [review]

**Review by hellesvik-nordic — APPROVED** [review]

**Comment by greg-fer** [comment]
@nordicjm , can you have another look? I've applied your edits. Thanks!

**Review by umapraseeda — APPROVED** [review]

File: doc/nrf/config_and_build/configuring_app/output_build_files.rst
Comment (umapraseeda) [inline doc/nrf/config_and_build/configuring_app/output_build_files.rst:1]: I think we would need to add this change also to redirects.py.

**Review by trond-snekvik — APPROVED** [review]
